pkg/api/credentials.go

```diff
@@ -50,43 +55,61 @@ func (h *CredentialsHandler) GenerateCredential(w http.ResponseWriter, r *http.R
 	}
 
 	// check if credential already exists
-	if exists, err := h.credRepo.DoesCredentialExist(req.HashedCredential); err != nil || exists {
+	if exists, err := h.credRepo.Exists(req.HashedCredential); err != nil || exists {
 		// TODO: better return 200 with message to fetch the wallet address or credential on a different endpoint
 		http.Error(w, "Credential already exists", http.StatusBadRequest)
 		return
 	}
 
-	// TODO: generate merkle tree root
-	// root, err := h.merkleTree.GenerateMerkleTree(req.HashedCredential)
-	// if err != nil {
-	// 	http.Error(w, "Failed to generate merkle tree root", http.StatusInternalServerError)
-	// 	return
-	// }
+	root, node, err := h.merkleTree.AddNodeToTree(req.WalletAddress, req.HashedCredential)
+	if err != nil {
+		http.Error(w, "Failed to generate merkle tree root", http.StatusInternalServerError)
+		return
+	}
 
-	// store merkle tree root
-	// store leaf in merkle tree
-	if err := h.credRepo.SaveCredential(req.HashedCredential); err != nil {
+	if err := h.credRepo.AddNodeToRoot(context.Background(), root.String(), node); err != nil {
 		http.Error(w, "Failed to save credential", http.StatusInternalServerError)
 		return
 	}
 
-	// generate wallet address
-	walletAddress, privKey, err := GenerateWalletAddress()
-	if err != nil {
-		http.Error(w, "Failed to generate wallet address", http.StatusInternalServerError)
-		return
+	walletAddress := ""
+
+	if len(strings.TrimSpace(req.WalletAddress)) == 0 {
```

Comment:
wallet is generated on the backend, that's the main idea to abstract it from the user.